packages/components/src/List/Toolbar/ColumnChooser/hooks/columnChooserManager.hook.js

```diff
@@ -0,0 +1,150 @@
+import { useState } from 'react';
+import cloneDeep from 'lodash/cloneDeep';
+import { compareOrder } from '../service';
+
+export function organiseColumns(columns) {
+	return columns.sort(compareOrder).map((column, index) => {
+		return { ...column, order: index + 1 };
+	});
+}
+
+export function updateEditedColumns(state) {
+	return { ...state, editedColumns: organiseColumns(state.editedColumns) };
+}
+
+export function isValueCorrect(value, collectionLength) {
+	return Number.isInteger(value) && value <= collectionLength;
+}
+
+export function getOrderItem(order, index, length) {
+	if (index === length - 1) {
+		return order - 1;
+	}
+	return order + 1;
+}
+
+export function matchOrder(value) {
+	return function match({ order }) {
+		return order === value;
+	};
+}
+
+export function changeColumnAttribute(key) {
+	return function setAttribut(value, index) {
+		return function setColumn(state) {
+			const newColumns = state.editedColumns;
+			newColumns[index][key] = value;
+			return { ...state, editedColumns: newColumns };
+		};
+	};
+}
+
+const updateAttributeVisiblity = changeColumnAttribute('hidden');
+const updateAttributeOrder = changeColumnAttribute('order');
+
+export function useColumnChooserManager(columns, customSubmit) {
+	const [state, setState] = useState({
+		editedColumns: organiseColumns(columns),
+		selectAll: false,
+	});
+
+	function getEditedColumnsLength() {
+		return state.editedColumns.length;
+	}
+
+	function getValueFormated(value) {
+		const parseValue = parseInt(value, 10);
+		if (isValueCorrect(parseValue, getEditedColumnsLength())) {
+			return parseValue;
+		}
+		throw Error(`ColumnChooserManager: Bad order number ${value}`);
+	}
+
+	function modifyOrderTwoItems(value, index) {
+		const indexColToReplace = state.editedColumns.findIndex(matchOrder(value));
+		if (indexColToReplace > -1 && !state.editedColumns[indexColToReplace].locked) {
+			setState(
+				updateAttributeOrder(
+					getOrderItem(value, indexColToReplace, getEditedColumnsLength()),
+					indexColToReplace,
+				),
+			);
+			setState(updateAttributeOrder(value, index));
+			setState(updateEditedColumns);
```

Comment:
It feel really strange to me calling multiple time setState with this attributes, i think i miss a thing

Reply:
It's a batch of setState.
React doesn't re render during batch, it will wait for the batch to complete.

Reply:
To change that I have to rework my set of  function.